### PR TITLE
mgr/DaemonServer: be more conservative when approving safe-to-destroy

### PR DESCRIPTION
**Pull request by xiexingguo**
by tracing back to any degraded PGs that ever stored on those
specific osds.

However, this is still not very accurate, e.g., we could:
- out an osd say osd.1, wait until all PGs become active+clean
- then out yet another osd (osd.2), check whether osd.1 is
  safe-to-destroy immediately

Step 2 will probably fail if osd.2 happens to share some PGs
(through crush) with osd.1.

In other words, we might become over-conservative when try to approve
a safe-to-destroy request, but I think it should be fine since we still
provide the "--force" option to allow user to arbitrarily destroy an osd?

Fixes: http://tracker.ceph.com/issues/38930
Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by xiexingguo** [comment]
jenkins test docs

**Comment by xiexingguo** [comment]
@liewegas Ping?

File: src/mgr/DaemonServer.cc [hunk 1]
```diff
@@ -1395,7 +1398,36 @@ bool DaemonServer::_handle_command(
...
+              auto new_weight = old_weight ? old_weight : CEPH_OSD_IN;
+              tmpmap.set_weight(osd, new_weight);
+              for (auto &i : pg_map.pg_stat) {
+                auto& pg = i.first;
+                auto state = i.second.state;
+                if (state & (PG_STATE_DEGRADED | PG_STATE_UNDERSIZED)) {
+                  vector<int> up;
+                  int up_primary;
+                  tmpmap.pg_to_raw_up(pg, &up, &up_primary);
+                  if (std::find(up.begin(), up.end(), osd) != up.end()) {
```
Comment (liewegas) [inline src/mgr/DaemonServer.cc]: I'm not sure checking whether this osd *would* store the PG if it were up+in tells us much.

What we really want to know is if there is any possible way that this OSD has data that we may need to recover.  I think the way to do that is to identify any non-active+clean PG, and see whether this OSD has a copy of that PG.  If it does, then we shouldn't destroy it.

Comment (xiexingguo) [reply src/mgr/DaemonServer.cc]: > I think the way to do that is to identify any non-active+clean PG, and see whether this OSD has a copy of that PG. If it does, then we shouldn't destroy it.

That's the problem. If that osd is currently down& out, we are going to erase any pg or osd related statistics from it. I can't think of a way of doing this without changing the existing stat-counting behaviour much.

**Comment by xiexingguo** [comment]
I don't think this is a ideal fix, closing :-(

**Pull request closed without merging**